src/Features/CSharp/Portable/Completion/CompletionProviders/PropertySubPatternCompletionProvider.cs

```diff
@@ -33,56 +34,112 @@ public PropertySubpatternCompletionProvider()
+            var type = GetMemberAccessType(propertyPatternType, memberAccess, document, semanticModel, position);
+
+            if (type is null)
             {
                 return;
             }
 
             // Find the members that can be tested.
-            IEnumerable<ISymbol> members = semanticModel.LookupSymbols(position, type);
-            members = members.Where(m => m.CanBeReferencedByName &&
-                IsFieldOrReadableProperty(m) &&
-                !m.IsImplicitlyDeclared &&
-                !m.IsStatic);
-
-            // Filter out those members that have already been typed
-            var propertyPatternClause = (PropertyPatternClauseSyntax)token.Parent;
+            var members = GetCandidatePropertiesAndFields(document, semanticModel, position, type);
 
-            // List the members that are already tested in this property sub-pattern
-            var alreadyTestedMembers = new HashSet<string>(propertyPatternClause.Subpatterns.Select(
-                // PROTOTYPE(extended-property-patterns) ExpressionColon
-                p => p.NameColon?.Name.Identifier.ValueText).Where(s => !string.IsNullOrEmpty(s)));
+            if (memberAccess is null)
+            {
+                // Filter out those members that have already been typed as simple (not extended) properties
+                var alreadyTestedMembers = new HashSet<string>(propertyPatternClause.Subpatterns.Select(
+                    p => p.NameColon?.Name.Identifier.ValueText).Where(s => !string.IsNullOrEmpty(s)));
 
-            var untestedMembers = members.Where(m => !alreadyTestedMembers.Contains(m.Name) &&
-                m.IsEditorBrowsable(document.ShouldHideAdvancedMembers(), semanticModel.Compilation));
+                members = members.WhereAsArray(m => !alreadyTestedMembers.Contains(m.Name));
+            }
 
-            foreach (var untestedMember in untestedMembers)
+            foreach (var member in members)
             {
                 const string ColonString = ":";
                 context.AddItem(SymbolCompletionItem.CreateWithSymbolId(
-                    displayText: untestedMember.Name.EscapeIdentifier(),
+                    displayText: member.Name.EscapeIdentifier(),
                     displayTextSuffix: ColonString,
                     insertionText: null,
-                    symbols: ImmutableArray.Create(untestedMember),
-                    contextPosition: token.GetLocation().SourceSpan.Start,
+                    symbols: ImmutableArray.Create(member),
+                    contextPosition: context.Position,
                     rules: s_rules));
             }
+
+            return;
+
+            // We have to figure out the type of the extended property ourselves, because
+            // the semantic model could not provide the answer we want in incomplete syntax:
+            // `c is { X. }`
```

Comment:
where are we landing with collection patterns?  do we have tests that this shows the right thing *when* 'X' binds to something in scope?

Reply:
List patterns are in progress in another feature branch (Ali is currently working on binding/lowering for those in PR https://github.com/dotnet/roslyn/pull/53299).
I did my best here to assume the two features will eventually meet, so hopefully merging should be smooth.